vmms/dockerSSH.py

```diff
@@ -0,0 +1,231 @@
+            if sys.platform == self._OS_X:
+                # boot2docker initialization will be part of initial
+                # set up with Tango.
+                self.docker_host_ip = subprocess.check_output(['boot2docker', 'ip']).strip('\n')
+            else:
+                self.docker_host_ip = self.LOCALHOST
+
+            self.log.info("Docker host IP is %s" % self.docker_host_ip)
+
+        except Exception as e:
+            self.log.error(str(e))
+            exit(1)
+
+    def instanceName(self, id, name):
+        """ instanceName - Constructs a VM instance name. Always use
+        this function when you need a VM instance name. Never generate
+        instance names manually.
+        """
+        # return str(id)
+        return "%s-%s-%s" % (config.Config.PREFIX, id, name)
+
+    def domainName(self, vm):
+        """ Returns the domain name that is stored in the vm
+        instance.
+        """
+        return vm.domain_name
+
+    #
+    # VMMS API functions
+    #
+    def initializeVM(self, vm):
+        """ initializeVM -  Nothing to do for initializeVM
+        """
+        return vm
+
+    def waitVM(self, vm, max_secs):
+        """ waitVM - Nothing to do for waitVM
+        """
+        return
+
+    def copyIn(self, vm, inputFiles):
+        """ copyIn - Create a directory to be mounted as a volume
+        for the docker containers. Copy input files to this directory.
+        """
+        instanceName = self.instanceName(vm.id, vm.name)
+
+        # Create a fresh volume
+        volume_path = config.Config.DOCKER_VOLUME_PATH + instanceName +'/'
+        os.makedirs(volume_path)
+        for file in inputFiles:
+            shutil.copy(file.localFile, volume_path + file.destFile)
+            self.log.info('Copied in file %s to %s' % (file.localFile, volume_path + file.destFile))
+        return 0
+
+    def runJob(self, vm, runTimeout, maxOutputFileSize):
+        """ runJob - Run a docker container by doing the follows:
+        - mount directory corresponding to this job to /home/autolab
+          in the container
+        - run autodriver with corresponding ulimits and timeout as
+          autolab user
+        """
+        instanceName = self.instanceName(vm.id, vm.name)
+        args = ['docker', 'run', '--name', instanceName, '-v']
+        args = args + ['%s:%s' % 
+                (config.Config.DOCKER_VOLUME_PATH + instanceName, '/home/mount')]
+        args = args + [config.Config.DOCKER_IMAGE]
+        args = args + ['sh', '-c']
+
+        autodriverCmd = 'autodriver -u %d -f %d -t %d -o %d autolab &> output/feedback' % \
+                        (config.Config.VM_ULIMIT_USER_PROC, 
+                        config.Config.VM_ULIMIT_FILE_SIZE,
+                        runTimeout, config.Config.MAX_OUTPUT_FILE_SIZE)
+
+        args = args + ['cp -r mount/* autolab/; su autolab -c "%s"; \
+                        cp output/feedback mount/feedback' % autodriverCmd]
```

Comment:
Right, so I was confused about that. And the difference between `CMD` and `ENTRYPOINT` is still not very clear to me. Shed some light on that? Also, the command is different for every job because the autodriver takes different timeout, ulimit and output file size for each job.